### ISSUE TITLE
Crash on vulkaninfo by loading libvulkan_radeon on an Intel GPU

### ISSUE THREAD

**Issue by thiagomacieira** [KhronosGroup/Vulkan-LoaderAndValidationLayers#1978]
When I run vulkaninfo, it crashes:
```
===========
VULKAN INFO
===========

Vulkan API Version: 1.0.41

INFO: [loader] Code 0 : Found manifest file /etc/vulkan/explicit_layer.d/VkLayer_core_validation.json, version "1.0.0"
INFO: [loader] Code 0 : Found manifest file /etc/vulkan/explicit_layer.d/VkLayer_object_tracker.json, version "1.0.0"
INFO: [loader] Code 0 : Found manifest file /etc/vulkan/explicit_layer.d/VkLayer_parameter_validation.json, version "1.0.0"
INFO: [loader] Code 0 : Found manifest file /etc/vulkan/explicit_layer.d/VkLayer_swapchain.json, version "1.0.0"
INFO: [loader] Code 0 : Found manifest file /etc/vulkan/explicit_layer.d/VkLayer_threading.json, version "1.0.0"
INFO: [loader] Code 0 : Found manifest file /etc/vulkan/explicit_layer.d/VkLayer_unique_objects.json, version "1.0.0"
INFO: [loader] Code 0 : Found manifest file /usr/share/vulkan/explicit_layer.d/VkLayer_core_validation.json, version "1.0.0"
INFO: [loader] Code 0 : Found manifest file /usr/share/vulkan/explicit_layer.d/VkLayer_object_tracker.json, version "1.0.0"
INFO: [loader] Code 0 : Found manifest file /usr/share/vulkan/explicit_layer.d/VkLayer_parameter_validation.json, version "1.0.0"
INFO: [loader] Code 0 : Found manifest file /usr/share/vulkan/explicit_layer.d/VkLayer_swapchain.json, version "1.0.0"
INFO: [loader] Code 0 : Found manifest file /usr/share/vulkan/explicit_layer.d/VkLayer_threading.json, version "1.0.0"
INFO: [loader] Code 0 : Found manifest file /usr/share/vulkan/explicit_layer.d/VkLayer_unique_objects.json, version "1.0.0"
INFO: [loader] Code 0 : Found ICD manifest file /usr/share/vulkan/icd.d/intel_icd.x86_64.json, version "1.0.0"
INFO: [loader] Code 0 : Found ICD manifest file /usr/share/vulkan/icd.d/radeon_icd.x86_64.json, version "1.0.0"

Program received signal SIGSEGV, Segmentation fault.
```
The backtrace points to libvulkan loading the Radeon driver:
```
(gdb) bt
#0  0x000000000001f436 in ?? ()
#1  0x00007fffede683f2 in ?? () from /usr/lib64/libvulkan_radeon.so
#2  0x00007ffff797f3e9 in loader_scanned_icd_add (api_version=4194307, 
    filename=0x7fffffffd0a0 "/usr/lib64/libvulkan_radeon.so", icd_tramp_list=0x5555557c2bb0, inst=0x5555557c2b70)
    at /usr/src/debug/Vulkan-LoaderAndValidationLayers-1.0.41/loader/loader.c:1565
#3  loader_icd_scan (inst=inst@entry=0x5555557c2b70, icd_tramp_list=icd_tramp_list@entry=0x5555557c2bb0)
    at /usr/src/debug/Vulkan-LoaderAndValidationLayers-1.0.41/loader/loader.c:3177
#4  0x00007ffff798506b in vkCreateInstance (pCreateInfo=0x7fffffffd660, pAllocator=0x0, pInstance=0x7fffffffd6b0)
    at /usr/src/debug/Vulkan-LoaderAndValidationLayers-1.0.41/loader/trampoline.c:329
#5  0x00005555555558c2 in AppCreateInstance (inst=0x7fffffffd6b0)
    at /usr/src/debug/Vulkan-LoaderAndValidationLayers-1.0.41/demos/vulkaninfo.c:667
#6  main (argc=<optimized out>, argv=<optimized out>)
    at /usr/src/debug/Vulkan-LoaderAndValidationLayers-1.0.41/demos/vulkaninfo.c:1454
```
My system has an Intel GPU:
```
$ lspci -s 0:2
00:02.0 VGA compatible controller: Intel Corporation Iris Graphics 540 (rev 0a)
```
The Intel driver is installed:
```
$ ls -l /usr/lib64/libvulkan*
-rwxr-xr-x 1 root root 2625088 jul 24 04:26 /usr/lib64/libvulkan_intel.so
-rwxr-xr-x 1 root root 1655912 jul 24 04:26 /usr/lib64/libvulkan_radeon.so
lrwxrwxrwx 1 root root      19 mai 27 22:52 /usr/lib64/libvulkan.so.1 -> libvulkan.so.1.0.41
-rwxr-xr-x 1 root root  239600 mai 27 22:52 /usr/lib64/libvulkan.so.1.0.41
```
Both files drivers are installed due to a dependency in package management, but clearly the Radeon one should not be used. The driver should validate that it is in the correct HW and bail out if not.

**Comment by thiagomacieira** [comment]
Also submitted at the radeon driver: https://bugs.freedesktop.org/show_bug.cgi?id=102010.

I don't know what is at fault: whether it's libvulkan for loading the wrong driver, or the driver for failing to properly bail out, or both for there not being a mechanism to bail out in the first place.

**Comment by thiagomacieira** [comment]
SUSE developers think the problem is actually a bug in glibc: the LLVM dependencies in the Radeon driver are causing the dynamic loader confusion.

https://bugzilla.opensuse.org/show_bug.cgi?id=1022156

**Comment by lenny-lunarg** [comment]
The main problem does appear to be in `libvulkan_radeon.so` (or the libraries that it loads). The loader has to load every ICD that is presented to it. Even if a driver doesn't have any physical devices associated with it, the loader can't know that until it loads the driver, so the problem here isn't from loading the wrong driver &mdash; that's intentional. If the driver can't be used on the system, it should just report no physical devices, and no one will ever be able to create a `VkDevice` with it.

The reason the loader keeps loading and unloading the ICD (and causing the crash after a couple tries) is that the loader isn't supposed to save any state until `vkCreateInstance` is called. This means that if the programmer calls anything before that (such as `vkEnumerateInstanceExtensionProperties`), the loader has to load all ICDs at the start of the call, and unload them at the end.

The only issue that I see with the loader is that it would be nice if the loader could detect the error, and return an error without crashing. I don't have easy access to a computer with `libvulkan_radeon.so` installed so I haven't been able to actually test this, but would you mind setting the envirment variable `LD_BIND_NOW` (setting it to 1 will work) and then running `vulkaninfo`? This environment variable will cause the loader to attempt to resolve all symbols when it loads the ICD, so it may fail more gracefully. We don't enable that behavior by default because there are negative performance implications associated with doing so.

**Comment by thiagomacieira** [comment]
`LD_BIND_NOW=1` does solve the problem. This does indeed look like a dynamic linker bug.

**Comment by lenny-lunarg** [comment]
Alright, then I'm going to close this issue. We've actually had this problem before, where a driver doesn't resolve it's symbols properly. We changed the loader to resolve all symbols at one point (commit a480e11deebf69a1b1dfebccd5314492409db601), but we had to revert that later (commit 18ef0ba5c47c214bb4c040c4ccaf9afd025fe25d) when it became apparent that we wanted the improved performance. As a result, I'm afraid we don't have a way to fail gracefully by default. I think we just have to live with the crash if the drivers are incorrect.
